include/deal.II/base/work_stream.h

```diff
@@ -874,14 +1019,352 @@ namespace WorkStream
+                  worker(i, scratch_data, copy_data);
+                if (static_cast<const std::function<void(const CopyData &)> &>(
+                      copier))
+                  copier(copy_data);
+              }
+
+            return;
+          }
+
+        tf::Executor &executor = MultithreadInfo::get_taskflow_executor();
+        tf::Taskflow  taskflow;
+
+        tf::Task last_copier = taskflow.placeholder();
+
+        Threads::ThreadLocalStorage<std::unique_ptr<ScratchData>>
+          thread_local_scratch;
+
+        std::vector<std::unique_ptr<Chunk<CopyData>>> chunks;
+
+        unsigned int idx             = 0;
+        unsigned int remaining_items = std::distance(begin, end);
+
+        const unsigned int real_chunk_size =
+          (remaining_items / chunk_size < 3 * MultithreadInfo::n_threads()) ?
+            1 :
+            chunk_size;
+
+        Iterator it = begin;
+        while (it != end)
+          {
+            unsigned int count  = std::min(remaining_items, real_chunk_size);
+            Iterator     middle = it;
+            std::advance(middle, count);
+
+            chunks.emplace_back();
+
+            // this chunk works on [it,middle)
+            auto worker_task =
+              taskflow
+                .emplace([it_begin = it,
+                          it_end   = middle,
+                          idx,
+                          count,
+                          &sample_scratch_data,
+                          &thread_local_scratch,
+                          &chunks,
+                          &sample_copy_data,
+                          &worker]() {
+                  auto &scratch_ptr = thread_local_scratch.get();
+                  if (!scratch_ptr.get())
+                    scratch_ptr =
+                      std::make_unique<ScratchData>(sample_scratch_data);
+
+                  ScratchData &scratch = *scratch_ptr.get();
+                  chunks[idx] =
+                    std::make_unique<Chunk<CopyData>>(count, sample_copy_data);
+
+                  unsigned int counter = 0;
+                  for (Iterator it = it_begin; it != it_end; ++it, ++counter)
+                    {
+                      worker(it,
+                             scratch,
+                             chunks[idx].get()->copy_datas[counter]);
+                    }
+                })
+                .name("work");
+
+            tf::Task copier_task =
+              taskflow
+                .emplace([idx, &chunks, &copier]() mutable {
+                  auto chunk = chunks[idx].get();
+                  for (auto &cd : chunk->copy_datas)
+                    copier(cd);
+
+                  chunks[idx].reset();
+                })
+                .name("copy");
+
+            worker_task.precede(copier_task);
+
+            last_copier.precede(copier_task);
+            last_copier = copier_task;
+
+            it = middle;
+            remaining_items -= count;
+            ++idx;
```

Comment:
You never set `chunks` to the right size. I think you're going to access invalid memory when you do `chunks[idx]` :-)

Your current design has each worker allocate a `Chunks` object, and the copier then deletes it again. That seems inefficient, along with the approach to have a vector that will have size `n_objects/chunk_size`. How did we solve that before, in the previous implementation?

Reply:
I think the code is not doing something invalid. 

My understanding is that the TBB filter has ``queue_length`` number of Chunks in flight in parallel. It seems that we re-use the copy objects as well, but I am not sure I understand it fully. I would think that the copy objects need to be reset, but I can't see that in the TBB implementation.
So yes, what I do might be less efficient. I am open to have a discussion about the efficiency and what other options we have. Maybe this is easier in a group call if you are interested? 
Alternatively, we start with one implementation and try to improve things over time.

Reply:
I looked into this again, and we have a buffer with as many chunk objects as we have tokens in flight. The copier simply marks a chunk object as unused again, and the token creator (the first stage of the pipeline) finds itself an unused chunk object in the buffer. (We call this the "ring buffer" in the implementation.) Because we know the maximum number of tokens in flight is at most as many as there are elements in the buffer, we know that when we need to find an unused one that there really is an unused one.

This points to one of the downsides of the design here, namely that we need to build the whole graph of tasks at once. The TBB design managed to build it on the fly, and that makes it substantially less memory intensive.